File: frontend/src/chapters/ch17/ch17_12_ex.ts
import { Lesson } from '../../courses';

export const ch17_12_ex: Lesson = {
    id: 'ch17-12-ex',
    title: 'Bài tập 17.12: Function with dyn Trait Parameter',
    duration: '15 phút',
    type: 'practice',
    isExercise: true,
    content: `<p>Hãy truyền dyn trait vào function!</p>`,
    defaultCode: `trait Describable { fn describe(&self) -> String; }
struct Book { title: String, pages: u32 }
impl Describable for Book { fn describe(&self) -> String { format!("{} - {} trang", self.title, self.pages) } }
fn print_description(item: &dyn Describable) { println!("{}", item.describe()); }
fn main() { let book = Book { title: String::from("Rust Programming"), pages: 350 }; print_description(&book); }
`,
    expectedOutput: 'Rust Programming',
    testCases: [{ input: 'describe()', expectedOutput: 'Rust Programming', description: 'dyn trait parameter works' }]
};

File: backend/src/exercises/ch17/ch17_12_ex.rs
// Exercise - ch17_12_ex: Function with dyn Trait Parameter

trait Describable {
    fn describe(&self) -> String;
}

struct Book {
    title: String,
    pages: u32,
}

struct Car {
    model: String,
    year: u32,
}

impl Describable for Book {
    fn describe(&self) -> String {
        format!("{} - {} trang", self.title, self.pages)
    }
}

impl Describable for Car {
    fn describe(&self) -> String {
        format!("{} - {}", self.model, self.year)
    }
}

fn print_description(item: &dyn Describable) {
    println!("{}", item.describe());
}

fn main() {
    let book = Book { title: String::from("Rust Programming"), pages: 350 };
    let car = Car { model: String::from("Toyota"), year: 2020 };

    print_description(&book);
    print_description(&car);
}
